src/typecheck.rs

```diff
@@ -1,172 +1,246 @@
 use std::ops::AddAssign;
```

Comment:
Can we pls rename this file to `query-compiler.rs` or something like this?
And move out a few of the typechecking related things to typechecking.rs

Reply:
Renamed to `core.rs`, moved some core-related code to this file, and moved Action executor and compiler to a separate `action.rs`.

Reply:
Amazing change, thank you!